src/layers/data_type_layer.cpp

```diff
@@ -282,45 +290,57 @@ auto data_type_layer<TensorDataType>::get_error_signals(int parent_index) const
 }
 
 // Accessing local matrices
-template <typename TensorDataType>
-auto data_type_layer<TensorDataType>::get_local_activations(int child_index) -> AbsMatrixType& {
+template <typename InputTensorDataType, typename OutputTensorDataType>
+auto data_type_layer<InputTensorDataType, OutputTensorDataType>::
+get_local_activations(int child_index) -> OutputAbsMatrixType& {
   return get_activations(child_index).Matrix();
 }
-template <typename TensorDataType>
-auto data_type_layer<TensorDataType>::get_local_error_signals(int parent_index) -> AbsMatrixType& {
+template <typename InputTensorDataType, typename OutputTensorDataType>
+auto data_type_layer<InputTensorDataType, OutputTensorDataType>::
+get_local_error_signals(int parent_index) -> InputAbsMatrixType& {
   return get_error_signals(parent_index).Matrix();
 }
-template <typename TensorDataType>
-auto data_type_layer<TensorDataType>::get_local_prev_activations(int parent_index) const -> const AbsMatrixType&{
+template <typename InputTensorDataType, typename OutputTensorDataType>
+auto data_type_layer<InputTensorDataType, OutputTensorDataType>::
+get_local_prev_activations(int parent_index) const -> const InputAbsMatrixType&{
   return get_prev_activations(parent_index).LockedMatrix();
 }
-template <typename TensorDataType>
-auto data_type_layer<TensorDataType>::get_local_activations(int child_index) const -> const AbsMatrixType& {
+template <typename InputTensorDataType, typename OutputTensorDataType>
+auto data_type_layer<InputTensorDataType, OutputTensorDataType>::
+get_local_activations(int child_index) const -> const OutputAbsMatrixType& {
   return get_activations(child_index).LockedMatrix();
 }
-template <typename TensorDataType>
-auto data_type_layer<TensorDataType>::get_local_prev_error_signals(int child_index) const -> const AbsMatrixType& {
+template <typename InputTensorDataType, typename OutputTensorDataType>
+auto data_type_layer<InputTensorDataType, OutputTensorDataType>::
+get_local_prev_error_signals(int child_index) const -> const OutputAbsMatrixType& {
   return get_prev_error_signals(child_index).LockedMatrix();
 }
-template <typename TensorDataType>
-auto data_type_layer<TensorDataType>::get_local_error_signals(int parent_index) const -> const AbsMatrixType& {
+template <typename InputTensorDataType, typename OutputTensorDataType>
+auto data_type_layer<InputTensorDataType, OutputTensorDataType>::
+get_local_error_signals(int parent_index) const -> const InputAbsMatrixType& {
   return get_error_signals(parent_index).LockedMatrix();
 }
 
 // Accessing matrices corresponding to parent/child layer
-template <typename TensorDataType>
-auto data_type_layer<TensorDataType>::get_activations(const Layer& child) const -> const BaseDistMat& {
+template <typename InputTensorDataType, typename OutputTensorDataType>
+auto data_type_layer<InputTensorDataType, OutputTensorDataType>::
+get_activations(const Layer& child) const -> const BaseDistMat& {
```

Comment:
Why is this returning a BaseDistMat not an Output matrix?

Reply:
Okay, I see that this is getting another layer's activations, so it may be a pain to figure out what it's output type, but shouldn't we know this statically?